bot/utils/redis_cache.py

```diff
@@ -0,0 +1,212 @@
+        """Turns all contents of a dict into typestrings."""
+        return {self._to_typestring(key): self._to_typestring(value) for key, value in dictionary.items()}
+
+    async def _validate_cache(self) -> None:
+        """Validate that the RedisCache is ready to be used."""
+        if self.bot is None:
+            raise RuntimeError("Critical error: RedisCache has no `Bot` instance.")
+
+        if self._namespace is None:
+            raise RuntimeError(
+                "Critical error: RedisCache has no namespace. "
+                "Did you initialize this object as a class attribute?"
+            )
+        await self.bot._redis_ready.wait()
+
+    def __set_name__(self, owner: Any, attribute_name: str) -> None:
+        """
+        Set the namespace to Class.attribute_name.
+
+        Called automatically when this class is constructed inside a class as an attribute.
+        """
+        self._set_namespace(f"{owner.__name__}.{attribute_name}")
+
+    def __get__(self, instance: RedisCache, owner: Any) -> RedisCache:
+        """Fetch the Bot instance, we need it for the redis pool."""
+        if self.bot:
+            return self
+
+        if self._namespace is None:
+            raise RuntimeError("RedisCache must be a class attribute.")
+
+        if instance is None:
+            raise RuntimeError("You must create an instance of RedisCache to use it.")
+
+        for attribute in vars(instance).values():
+            if isinstance(attribute, Bot):
+                self.bot = attribute
+                self._redis = self.bot.redis_session
+                return self
+        else:
+            raise RuntimeError("Cannot initialize a RedisCache without a `Bot` instance.")
+
+    def __repr__(self) -> str:
+        """Return a beautiful representation of this object instance."""
+        return f"RedisCache(namespace={self._namespace!r})"
+
+    async def set(self, key: ValidRedisType, value: ValidRedisType) -> None:
+        """Store an item in the Redis cache."""
+        await self._validate_cache()
+
+        # Convert to a typestring and then set it
+        key = self._to_typestring(key)
+        value = self._to_typestring(value)
+        await self._redis.hset(self._namespace, key, value)
+
+    async def get(self, key: ValidRedisType, default: Optional[ValidRedisType] = None) -> ValidRedisType:
+        """Get an item from the Redis cache."""
+        await self._validate_cache()
+        key = self._to_typestring(key)
+        value = await self._redis.hget(self._namespace, key)
+
+        if value is None:
+            return default
+        else:
+            value = self._from_typestring(value)
+            return value
+
+    async def delete(self, key: ValidRedisType) -> None:
+        """Delete an item from the Redis cache."""
+        await self._validate_cache()
+        key = self._to_typestring(key)
+        return await self._redis.hdel(self._namespace, key)
```

Comment:
We rely on the "ignore missing keys"-behavior HDEL has elsewhere (in `.pop`). I think we should document the behavior of this method when confronted with missing keys here.

Reply:
I've added some documentation for this, but I also decided to not rely on it for `.pop`, since it's just an additional superfluous API call for no real benefit.